### PR TITLE
[Hotfix] Redirect `/download` [OSF-7835]

### PR DESCRIPTION
**Pull request by mfraezz**
## Purpose
Redirect `/download` for files and preprints

## Changes
Add special cases in `resolve_guid`

## Side effects
None expected

## Ticket
[[OSF-7835]](https://openscience.atlassian.net/browse/OSF-7835)

### CODE REVIEW

File: osf_tests/test_guid.py [hunk 1]
```diff
@@ -160,3 +160,19 @@ def test_resolve_guid_no_url(self):
...
+        pp = PreprintFactory(finish=True)
+        res = self.app.get(pp.url + 'download')
+        assert res.status_code == 302
+        assert res.location == 'http://localhost:7777/v1/resources/{}/providers/osfstorage/{}'.format(pp.node._id, pp.primary_file._id)
+
+    def test_resolve_guid_preprint_download_export(self):
+        pp = PreprintFactory(finish=True)
+        res = self.app.get(pp.url + 'download?format=asdf')
+        assert res.status_code == 302
+        assert res.location == 'http://localhost:7778/export?format=asdf&url=http://localhost:7777/v1/resources/{}/providers/osfstorage/{}'.format(pp.node._id, pp.primary_file._id)
```
Comment (sloria) [inline osf_tests/test_guid.py]: Would be good to use `MFR_SERVER_URL` here.

File: osf_tests/test_guid.py [hunk 2]
```diff
@@ -160,3 +160,19 @@ def test_resolve_guid_no_url(self):
             expect_errors=True,
         )
         assert res.status_code == 404
+
+    def test_resolve_guid_preprint_download(self):
+        pp = PreprintFactory(finish=True)
+        res = self.app.get(pp.url + 'download')
+        assert res.status_code == 302
+        assert res.location == 'http://localhost:7777/v1/resources/{}/providers/osfstorage/{}'.format(pp.node._id, pp.primary_file._id)
```
Comment (sloria) [inline osf_tests/test_guid.py]: Use `WATERBUTLER_URL`.

File: website/views.py [hunk 3]
```diff
@@ -241,6 +242,10 @@ def resolve_guid(guid, suffix=None):
         if not referent.deep_url:
             raise HTTPError(http.NOT_FOUND)
         if isinstance(referent, PreprintService):
+            if re.match('download/?', suffix) and referent.primary_file:
```
Comment (icereval) [inline website/views.py]: I'm not certain about the beginning and ending slashes - regex seems to be overkill... and is slower

`suffix.starts_with('/download')`

Comment (icereval) [inline website/views.py]: BTW, this also needs to work for file guids

File: website/views.py [hunk 4]
```diff
@@ -241,6 +242,10 @@ def resolve_guid(guid, suffix=None):
...
+                if request.args.get('format'):
+                    return redirect('{}/export?format={}&url={}'.format(MFR_SERVER_URL, request.args['format'], referent.primary_file.generate_waterbutler_url()))
```
Comment (icereval) [inline website/views.py]: The URL needs to use the `urllib.url_encode`? equivalent of JS' `encodeURIComponent`

File: website/views.py (hunk 3 above)
Comment (mfraezz) [inline website/views.py]: No beginning slash, possible ending slash -- `starts_with` should be fine

**Comment by mfraezz** [comment]
Build failure because upstream is failing, unrelated to these changes

**Pull request merged**